crates/polars-utils/src/order_statistic_tree.rs

```diff
@@ -0,0 +1,705 @@
+        let a = self.tree_unique_weight(left);
+        let b = self.tree_unique_weight(right);
+        a + 1 < GAMMA * (b + 1)
+    }
+
+    #[inline]
+    pub fn rank_lower(&self, bound: &T) -> Result<usize, usize> {
+        self.rank_lower_inner(bound, self.root)
+    }
+
+    fn rank_lower_inner(&self, value: &T, tree: TreeKey) -> Result<usize, usize> {
+        if tree.is_null() {
+            return Err(0);
+        }
+        let tn = &self.tree_nodes[tree];
+        let vn = &self.value_nodes[tn.values.head];
+        match (self.compare)(value, &vn.value) {
+            Ordering::Less => self.rank_lower_inner(value, tn.left),
+            Ordering::Equal => Ok(self.tree_weight(tn.left)),
+            Ordering::Greater => self
+                .rank_lower_inner(value, tn.right)
+                .map(|rank| self.tree_weight(tree) - self.tree_weight(tn.right) + rank)
+                .map_err(|rank| self.tree_weight(tree) - self.tree_weight(tn.right) + rank),
+        }
+    }
+
+    #[inline]
+    pub fn rank_upper(&self, bound: &T) -> Result<usize, usize> {
+        self._rank_upper(bound, self.root)
+    }
+
+    fn _rank_upper(&self, value: &T, tree: TreeKey) -> Result<usize, usize> {
+        if tree.is_null() {
+            return Err(self.tree_weight(tree));
+        }
+        let tn = &self.tree_nodes[tree];
+        let vn = &self.value_nodes[tn.values.head];
+        match (self.compare)(value, &vn.value) {
+            Ordering::Less => self._rank_upper(value, tn.left),
+            Ordering::Equal => Ok(self.tree_weight(tree) - self.tree_weight(tn.right) - 1),
+            Ordering::Greater => self
+                ._rank_upper(value, tn.right)
+                .map(|rank| self.tree_weight(tree) - self.tree_weight(tn.right) + rank)
+                .map_err(|rank| self.tree_weight(tree) - self.tree_weight(tn.right) + rank),
+        }
+    }
+
+    #[inline]
+    pub fn rank_unique(&self, value: &T) -> Result<usize, usize> {
+        self.rank_unique_inner(value, self.root)
+    }
+
+    fn rank_unique_inner(&self, value: &T, tree: TreeKey) -> Result<usize, usize> {
+        if tree.is_null() {
+            return Err(0);
+        }
+        let tn = &self.tree_nodes[tree];
+        let vn = &self.value_nodes[tn.values.head];
+        match (self.compare)(value, &vn.value) {
+            Ordering::Less => self.rank_unique_inner(value, tn.left),
+            Ordering::Equal => Ok(self.tree_unique_weight(tn.left)),
+            Ordering::Greater => self
+                .rank_unique_inner(value, tn.right)
+                .map(|rank| {
+                    self.tree_unique_weight(tree) - self.tree_unique_weight(tn.right) + rank
+                })
+                .map_err(|rank| {
+                    self.tree_unique_weight(tree) - self.tree_unique_weight(tn.right) + rank
+                }),
+        }
+    }
+
+    #[inline]
+    pub fn count(&self, value: &T) -> usize {
```

Comment:
Can't this simply be a search for `value` returning `values.len` if we find an exact match and 0 otherwise?